File: javascript/circular-buffer/circular-buffer.js
class CircularBuffer {
    constructor(size) {
        this.buffer = new Array(size);
        this.head = 0;
        this.tail = 0;
    }

    get size() {
        return this.tail - this.head;
    }

    read() {
        if (this.size == 0)
            throw new BufferEmptyError;

        return this.buffer[this.head++ % this.buffer.length];
    }

    write(x) {
        if (this.size == this.buffer.length)
            throw new BufferFullError;

        if (x) {
            this.buffer[this.tail % this.buffer.length] = x;
            this.tail += 1;
        }
    }

    forceWrite(x) {
        try {
            this.write(x);
        } catch(err) {
            if (err instanceof BufferFullError) {
                this.buffer[this.tail % this.buffer.length] = x;

                this.head += 1;
                this.tail += 1;
            } else {
                throw err;
            }
        }
    }

    clear() {
        this.head = this.tail = 0;
    }
}

export default CircularBuffer;

export class BufferFullError extends Error {}

export class BufferEmptyError extends Error {}

File: c/circular-buffer/src/circular_buffer.c
#include "circular_buffer.h"

circular_buffer_t* new_circular_buffer(size_t capacity) {
    circular_buffer_t* c = malloc(sizeof(circular_buffer_t));

    *c = (circular_buffer_t){
        .head = 0, .tail = 0,
        .capacity = capacity,
        .buffer = malloc(sizeof(int16_t) * capacity)
    };

    return c;
}

size_t size(circular_buffer_t* c) {
    return c->tail - c->head;
}

int16_t read(circular_buffer_t* c, int16_t* out) {
    if (size(c) == 0) {
        errno = ENODATA;
        return EXIT_FAILURE;
    }

    *out = c->buffer[c->head++ % c->capacity];

    return EXIT_SUCCESS;
}

int16_t write(circular_buffer_t* c, int16_t value) {
    if (size(c) == c->capacity) {
        errno = ENOBUFS;
        return EXIT_FAILURE;
    }

    c->buffer[c->tail++ % c->capacity] = value;

    return EXIT_SUCCESS;
}

int16_t overwrite(circular_buffer_t* c, int16_t value) {
    write(c, value);

    if (errno == ENOBUFS) {
        c->buffer[c->tail % c->capacity] = value;

        c->head += 1;
        c->tail += 1;
    }

    return EXIT_SUCCESS;
}

void delete_buffer(circular_buffer_t* c) {
    free(c->buffer);
    free(c);
}

void clear_buffer(circular_buffer_t* c) {
    c->head = 0;
    c->tail = 0;
}
